File: packages/sdk-typescript/src/core/retries.ts
import {
  ApiError,
  BimpeAIError,
  ConflictError,
  ConnectionError,
  ConnectionTimeoutError,
  NotImplementedError,
} from './errors';

export const DEFAULT_BASE_MS = 500;
export const DEFAULT_MAX_BACKOFF_MS = 8000;

const RETRYABLE_STATUSES = new Set([408, 429]);

export function shouldRetry(error: unknown, attempt: number, maxRetries: number): boolean {
  if (attempt >= maxRetries) return false;
  if (!(error instanceof BimpeAIError)) return false;
  if (error instanceof ConnectionTimeoutError || error instanceof ConnectionError) return true;
  if (error instanceof NotImplementedError) return false;
  if (error instanceof ConflictError) return false;
  if (error instanceof ApiError) {
    if (RETRYABLE_STATUSES.has(error.status)) return true;
    if (error.status >= 500 && error.status !== 501) return true;
  }
  return false;
}

export function computeBackoff(
  attempt: number,
  baseMs: number = DEFAULT_BASE_MS,
  maxMs: number = DEFAULT_MAX_BACKOFF_MS,
  retryAfterMs?: number,
): number {
  if (typeof retryAfterMs === 'number' && retryAfterMs >= 0) {
    return Math.min(retryAfterMs, maxMs);
  }
  const exponential = Math.min(maxMs, baseMs * 2 ** attempt);
  const jitter = 0.5 + Math.random() * 0.5;
  return Math.floor(exponential * jitter);
}

File: packages/sdk-python/src/bimpeai/_retries.py
from __future__ import annotations

import random

from ._exceptions import (
    APIConnectionError,
    APIError,
    APINotImplementedError,
    BimpeAIError,
    ConflictError,
)

DEFAULT_BASE_S = 0.5
DEFAULT_MAX_BACKOFF_S = 8.0

_RETRYABLE_STATUSES = {408, 429}


def should_retry(error: object, attempt: int, max_retries: int) -> bool:
    if attempt >= max_retries:
        return False
    if not isinstance(error, BimpeAIError):
        return False
    if isinstance(error, APIConnectionError):
        return True
    if isinstance(error, (APINotImplementedError, ConflictError)):
        return False
    if isinstance(error, APIError):
        if error.status in _RETRYABLE_STATUSES:
            return True
        if error.status >= 500 and error.status != 501:
            return True
    return False


def compute_backoff(
    attempt: int,
    base_s: float = DEFAULT_BASE_S,
    max_s: float = DEFAULT_MAX_BACKOFF_S,
    *,
    retry_after_s: float | None = None,
) -> float:
    if retry_after_s is not None and retry_after_s >= 0:
        return min(retry_after_s, max_s)
    exponential = min(max_s, base_s * (2**attempt))
    return exponential * (0.5 + random.random() * 0.5)
